File: exam-conductor/archiveDCR/libs/exampen-proto/python/exampen_proto/bff.py
"""BFF view models: teacher and student aggregation surfaces."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from .enums import ObjectionStatus, PassFail, StudentExamStatus


class TeacherExamCard(BaseModel):
    """Exam card for the teacher dashboard list."""

    exam_id: UUID
    title: str
    subject_id: str
    scheduled_at: datetime
    state: str
    class_label: Optional[str] = None


class ClassScoreRow(BaseModel):
    """One row in the teacher class score grid."""

    student_id: str
    student_name: str
    total_score: float
    percentile: Optional[float] = None
    ai_confidence: float
    miss_indicator_count: Optional[int] = None
    plagiarism_flag_count: Optional[int] = None


class QuestionDetail(BaseModel):
    """Per-question detail for teacher drill-down."""

    question_id: str
    current_score: float
    confidence: float
    recognized_text: Optional[str] = None
    miss_indicator: Optional[str] = None
    copy_image_uri: Optional[str] = None


class TeacherStudentDetail(BaseModel):
    """Teacher drill-down view for one student's exam."""

    student_id: str
    student_name: str
    total_score: float
    answer_pages: Optional[list[str]] = None
    questions: list[QuestionDetail]


class TeacherScoreOverrideRequest(BaseModel):
    """Teacher-initiated score override forwarded through BFF."""

    question_id: str
    new_score: float
    reason: str


class PlagiarismPreview(BaseModel):
    """Plagiarism flag preview for teacher review."""

    flag_id: UUID
    student_a_id: str
    student_b_id: str
    question_id: str
    composite_score: float
    severity: str
    teacher_verdict: Optional[str] = None


class ObjectionInboxItem(BaseModel):
    """Objection summary for the teacher inbox."""

    objection_id: UUID
    student_id: str
    question_id: str
    status: str
    filed_at: datetime


class StudentExamCard(BaseModel):
    """Exam card for the student portal list."""

    exam_id: UUID
    title: str
    subject_name: Optional[str] = None
    scheduled_at: datetime
    status: StudentExamStatus


class StudentQuestionScore(BaseModel):
    """Per-question score in the student score view."""

    question_id: str
    marks_obtained: float
    max_marks: float
    ai_confidence: Optional[float] = None
    miss_indicator: Optional[str] = None


class StudentScoreView(BaseModel):
    """Student-facing score summary for an exam."""

    exam_id: UUID
    total_score: float
    percentage: float
    percentile: float
    pass_fail: Optional[PassFail] = None
    questions: list[StudentQuestionScore]


class StudentObjection(BaseModel):
    """Student-facing objection record."""

    objection_id: UUID
    exam_id: UUID
    question_id: str
    status: ObjectionStatus
    objection_text: Optional[str] = None
    resolution_reason: Optional[str] = None
    new_score: Optional[float] = None


class CreateStudentObjectionRequest(BaseModel):
    """Student-initiated objection request."""

    exam_id: UUID
    question_id: str
    objection_text: str

File: exam-conductor/archiveDCR/libs/exampen-proto/typescript/src/bff.ts
/** BFF view types: teacher and student aggregation surfaces. */

import type { ObjectionStatus, PassFail, StudentExamStatus } from "./enums";

/** Exam card for the teacher dashboard list. */
export interface TeacherExamCard {
  exam_id: string;
  title: string;
  subject_id: string;
  scheduled_at: string;
  state: string;
  class_label?: string;
}

/** One row in the teacher class score grid. */
export interface ClassScoreRow {
  student_id: string;
  student_name: string;
  total_score: number;
  percentile?: number;
  ai_confidence: number;
  miss_indicator_count?: number;
  plagiarism_flag_count?: number;
}

/** Per-question detail for teacher drill-down. */
export interface QuestionDetail {
  question_id: string;
  current_score: number;
  confidence: number;
  recognized_text?: string;
  miss_indicator?: string;
  copy_image_uri?: string;
}

/** Teacher drill-down view for one student's exam. */
export interface TeacherStudentDetail {
  student_id: string;
  student_name: string;
  total_score: number;
  answer_pages?: string[];
  questions: QuestionDetail[];
}

/** Teacher-initiated score override forwarded through BFF. */
export interface TeacherScoreOverrideRequest {
  question_id: string;
  new_score: number;
  reason: string;
}

/** Plagiarism flag preview for teacher review. */
export interface PlagiarismPreview {
  flag_id: string;
  student_a_id: string;
  student_b_id: string;
  question_id: string;
  composite_score: number;
  severity: string;
  teacher_verdict?: string;
}

/** Objection summary for the teacher inbox. */
export interface ObjectionInboxItem {
  objection_id: string;
  student_id: string;
  question_id: string;
  status: string;
  filed_at: string;
}

/** Exam card for the student portal list. */
export interface StudentExamCard {
  exam_id: string;
  title: string;
  subject_name?: string;
  scheduled_at: string;
  status: StudentExamStatus;
}

/** Per-question score in the student score view. */
export interface StudentQuestionScore {
  question_id: string;
  marks_obtained: number;
  max_marks: number;
  ai_confidence?: number;
  miss_indicator?: string;
}

/** Student-facing score summary for an exam. */
export interface StudentScoreView {
  exam_id: string;
  total_score: number;
  percentage: number;
  percentile: number;
  pass_fail?: PassFail;
  questions: StudentQuestionScore[];
}

/** Student-facing objection record. */
export interface StudentObjection {
  objection_id: string;
  exam_id: string;
  question_id: string;
  status: ObjectionStatus;
  objection_text?: string;
  resolution_reason?: string;
  new_score?: number;
}

/** Student-initiated objection request. */
export interface CreateStudentObjectionRequest {
  exam_id: string;
  question_id: string;
  objection_text: string;
}
